proposals/BVAL-508.md

```diff
@@ -0,0 +1,572 @@
+---
+title: Validating elements contained in a container (like collections)
+layout: default
+author: Emmanuel Bernard
+comments: true
+---
+
+# #{page.title}
+
+[Link to JIRA ticket][jira]  
+[Related JIRA](https://hibernate.atlassian.net/browse/BVAL-499)  
+[Java generics glossary](/glossary/)  
+
+## Problem
+
+It is useful to be able to apply constrains on elements contained in a so called _container_.
+Examples of containers are:
+
+* `Collection`
+* `Optional`
+* JavaFX's `Property`
+
+But today the constraints are applied on the container value itself and not its content.
+
+## Proposition 1
+
+### Containers
+
+Bean Validation will offer the notion of container of values to validate.
+Containers can contain one value, or several values (`Optional` vs `Collection`).
+A container can contain several types of values: a `Map` contains keys and values for example.
+
+Extractors for the specific tuple container and value type can be implemented.
+
+TODO: should we differentiate single and plural?
```

Comment:
I think it's good for the sake of perf - no Iterable allocation in the singular case.